File: Binary_Search/BOJ_2805.py
# 0부터 나무 길이 최대값 까지의 범위를 이진 탐색하며 절단기 높이의 최대값을 구한다.
n, m = map(int, input().split())

trees = list(map(int, input().split()))

end = max(trees)
start = 0
result = 0

while(start <= end):
    sum = 0
    mid = (start + end) // 2
    for tree in trees:
        if tree > mid:
            sum += tree - mid
    if sum >= m:
        result = mid
        start = mid + 1
    # sum이 m보다 크면 sum을 줄이려는 시도 한다.(높이의 최댓값을 구한다는 의미는 sum의 최소값을 구하는 것)
    # sum을 줄이려면 mid가 커져야 하므로 start = mid + 1을 해준다.
    else:
        end = mid - 1

print(result)

File: C++/BinarySearch/BOJ_2805.cpp
/*
binary search, 최댓값 찾을때 까지 갱신
ios_base::sync_with_stdio(false);
cin.tie(0);
cout.tie(0);
위 세 줄 추가해서 입출력 속도 향상 (입력 개수 많을 때 시간 차이 많이남)
*/

#define ri register int
#define ll long long int
#include <algorithm>
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

int n, m;
vector<int> a;

int main(void) {
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	cin >> n >> m;

	for (ri i = 0; i < n; ++i) {
		int input;
		cin >> input;
		a.push_back(input);
	}

	int start = 0;
	int end = *max_element(a.begin(), a.end());
	int res = 0;

	while (start <= end) {
		int mid = (start + end) / 2;
		ll sum = 0;
		for (auto i : a) {
			if ((i - mid) > 0) sum += (i - mid);
		}
		if (sum < m) {
			end = mid - 1;
		}
		else {
			res = mid;
			start = mid + 1;
		}
	}

	cout << res;
	

	return 0;
}